File: takkernel/engine/src/main/cpp/feature/FeatureSet.cpp
#include "feature/FeatureSet.h"

using namespace atakmap::feature;

#define MEM_FN( fn )    "atakmap::feature::FeatureSet::" fn ": "

FeatureSet::FeatureSet (int64_t id,
                        unsigned long version,
                        const char* provider,
                        const char* type,
                        const char* name,
                        double minResolution,
                        double maxResolution)
    : ID(id),
      version(version),
      owner(nullptr),
      provider(provider),
      type(type),
      name(name),
      minResolution(minResolution),
      maxResolution(maxResolution)
{}

FeatureSet::FeatureSet (const char* provider,
                        const char* type,
                        const char* name,
                        double minResolution,
                        double maxResolution)
  : ID (FeatureDataStore::ID_NONE),
    version (FeatureDataStore::VERSION_NONE),
    owner (nullptr),
    provider (provider),
    type (type),
    name (name),
    minResolution (minResolution),
    maxResolution (maxResolution)
  {
    if (!provider)
      {
        throw std::invalid_argument (MEM_FN ("FeatureSet")
                                     "Received NULL provider");
      }
    if (!type)
      {
        throw std::invalid_argument (MEM_FN ("FeatureSet")
                                     "Received NULL type");
      }
    if (!name)
      {
        throw std::invalid_argument (MEM_FN ("FeatureSet")
                                     "Received NULL name");
      }
    if (minResolution < 0)
      {
        throw std::invalid_argument (MEM_FN ("FeatureSet")
                                     "Received negative minResolution");
      }
    if (maxResolution < 0)
      {
        throw std::invalid_argument (MEM_FN ("FeatureSet")
                                     "Received negative maxResolution");
      }
  }

FeatureSet::~FeatureSet ()
    NOTHROWS
    { }


FeatureDataStore::FeatureCursor*
FeatureSet::getFeatures ()
    const
    NOTHROWS
try
  {
    typedef std::shared_ptr<QueryParams::Order>   OrderPtr;

    if (!queryParams)
      {
        queryParams = QueryParamsPtr (new QueryParams ());
        queryParams->featureSetIDs = std::vector<int64_t> (1, ID);
        queryParams->orders = std::vector<OrderPtr>
                                  (1, OrderPtr (new QueryParams::Name ()));
      }

    return owner ? owner->queryFeatures (*queryParams) : nullptr;
  }
catch (...)
  { return nullptr; }


int64_t
FeatureSet::getID ()
    const
    NOTHROWS
    { return ID; }

//
// Returns the ground sample distance (in meters/pixel) of the "highest
// resolution" at which the features should be displayed.  A value of 0.0
// indicates that there is no maximum.
//
// N.B.:    As "resolution" increases (in the conventional sense), the
//          number of meters/pixel decreases; thus the value returned by
//          getMaxResolution will be less than or equal to the value
//          returned by getMinResolution.
//
double
FeatureSet::getMaxResolution ()
    const
    NOTHROWS
    { return maxResolution; }

//
// Returns the ground sample distance (in meters/pixel) of the "lowest
// resolution" at which the features should be displayed.  A value of 0.0
// indicates that there is no minimum.
//
// N.B.:    As "resolution" decreases (in the conventional sense), the
//          number of meters/pixel increases; thus the value returned by
//          getMinResolution will be greater than or equal to the value
//          returned by getMaxResolution.
//
double
FeatureSet::getMinResolution ()
    const
    NOTHROWS
    { return minResolution; }

const char*
FeatureSet::getName ()
    const
    NOTHROWS
    { return name; }

const char*
FeatureSet::getProvider ()
    const
    NOTHROWS
    { return provider; }

const char*
FeatureSet::getType ()
    const
    NOTHROWS
    { return type; }

unsigned long
FeatureSet::getVersion ()
    const
    NOTHROWS
    { return version; }

#undef MEM_FN

File: takkernel/engine/src/main/java/com/atakmap/map/layer/feature/FeatureSet.java
package com.atakmap.map.layer.feature;

import java.util.Collections;

import com.atakmap.annotations.DeprecatedApi;
import com.atakmap.map.layer.feature.cursor.MultiplexingFeatureCursor;

/**
 * A grouping of map features.
 * 
 * @author Developer
 */
public final class FeatureSet {

    FeatureDataStore owner;
    long id;
    long version;
    private final String provider;
    private final String type;
    private final String name;
    private final double minGsd;
    private final double maxGsd;
    private FeatureDataStore.FeatureQueryParameters featuresParam;
    
    public FeatureSet(String provider, String type, String name, double minGsd, double maxGsd) {
        this(null, FeatureDataStore.FEATURESET_ID_NONE, provider, type, name, minGsd, maxGsd, FeatureDataStore.FEATURESET_VERSION_NONE);
    }

    public FeatureSet(long id, String provider, String type, String name, double minGsd, double maxGsd, long version) {
        this(null, id, provider, type, name, minGsd, maxGsd, version);
    }

    FeatureSet(FeatureDataStore owner, long id, String provider, String type, String name, double minGsd, double maxGsd, long version) {
        this.owner = owner;
        this.id = id;
        this.provider = provider;
        this.type = type;
        this.name = name;
        this.minGsd = minGsd;
        this.maxGsd = maxGsd;

        this.featuresParam = null;
    }
    
    public String getProvider() {
        return this.provider;
    }

    public String getType() {
        return this.type;
    }

    /**
     * Returns the name of the feature set.
     * 
     * @return  The name of the feature set.
     */
    public String getName() {
        return this.name;
    }
    
    /**
     * Returns the ID of the feature set.
     * 
     * @return  The ID of the feature set.
     */
    public long getId() {
        return this.id;
    }
    
    /**
     * Returns the minimum display resolution of the feature set, in meters per
     * pixel. Larger values equate to lower resolutions.
     * 
     * @return  The minimum display resolution of the feature set.
     */
    public double getMinResolution() {
        return this.minGsd;
    }
    
    /**
     * Returns the maximum display resolution of the feature set, in meters per
     * pixel. Smaller values equate to higher resolutions.
     * 
     * @return  The maximum display resolution of the feature set.
     */
    public double getMaxResolution() {
        return this.maxGsd;
    }

    public long getVersion() {
        return this.version;
    }
}
